### PR TITLE
feat: display snap account name

### LINKED ISSUE #20819: Add snap label to Account Snaps in the account list
The snap label is missing in the accounts list. Account snaps should present a snap, including the snap badge and the name of the snap it's derived from.

Current displayed only as "snaps"
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBbXFKIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--38f6cbc82cd489f34f79b40142e52ce4226aa5e6/image.png)

Snapl label
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBbWlKIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--690d6858155a3452f9b15b2ac6f919d44134d334/image.png)

Figma file
https://www.figma.com/file/4jYXYiXhJfr1iknC1ElGVt/Account-details-%23446?type=design&node-id=6640%3A78639&mode=design&t=rXqNkhSfsTkqwX5s-1

### PR DESCRIPTION
**Pull request by k-g-j**
## **Description**

This feature adds a label for snap accounts with the derived snap name and MetaMask snap icon to the tag on the accounts list and wallet overview UI.

Reason: Account snaps should present a snap tag, including the snap badge and the name of the snap it's derived from.

Improvement: The account-list-menu has been updated to pass a keyring type and optional tag label to child account-list-item components. The wallet-overview screen now shows a tag with the snap name for snap accounts.

## **Related issues**

Fixes: #20819 

## **Manual testing steps**

1. Go to the [SSK](https://metamask.github.io/snap-simple-keyring/latest/) and create a snap account
2. Open the accounts list dropdown in the extension
3. You should see "MetaMask Simple Snap Keyring (Beta)" under the new account with the snap icon
4. Open the expanded view of the extension
5. You should see "MetaMask Simple Snap Keyring (Beta)" under the new account with the snap icon when you navigate to the new account page 
6. You should NOT see any tag under an account that is not a snap account (default MM accounts)

### **Jest**
`jest ui/components/multichain/account-list-menu/account-list-menu.test.js`
`jest ui/components/multichain/account-list-item/account-list-item.test.js`

### **Before**

Screenshot 2023-12-14 at 2 25 55 PM

Screenshot 2023-12-14 at 2 26 34 PM

### **After**

Screenshot 2023-12-14 at 2 29 24 PM

Screenshot 2023-12-14 at 2 30 04 PM

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/components/multichain/account-list-menu/account-list-menu.js [hunk 1]
```diff
@@ -64,6 +66,48 @@ const ACTION_MODES = {
...
+    case KeyringType.snap:
+      if (account.metadata.snap?.name) {
+        return `${account.metadata.snap?.name} (${translate('beta')})`;
+      }
+      return `${translate('snaps')} (${translate('beta')})`;
+    ///: END:ONLY_INCLUDE_IF
+    default:
+      return null;
+  }
+}
```
Comment (gantunesr) [inline ui/components/multichain/account-list-menu/account-list-menu.js:91]: I think you should move this method to an utils file instead of having it in a component and then just import it in the appropriate views/components

File: ui/components/app/wallet-overview/wallet-overview.js [hunk 2]
```diff
@@ -6,15 +6,24 @@ import { toChecksumHexAddress } from '../../../../shared/modules/hexstring-utils
...
   className,
   showAddress = false,
 }) => {
+  const t = useI18nContext();
   const selectedAccount = useSelector(getSelectedInternalAccount);
   const checksummedAddress = toChecksumHexAddress(selectedAccount.address);
+  const label = selectedAccount.metadata.snap?.name
+    ? `${selectedAccount.metadata.snap.name} (${t('beta')})`
+    : null;
+  const keyring = selectedAccount.metadata.keyring.type;
```
Comment (gantunesr) [inline ui/components/app/wallet-overview/wallet-overview.js:26]: Here for example, we can use the method `getLabel`

File: ui/components/multichain/account-list-menu/account-list-menu.js (hunk 1 above)
Comment (k-g-j) [reply ui/components/multichain/account-list-menu/account-list-menu.js:91]: Good Idea, done!

File: ui/components/app/wallet-overview/wallet-overview.js (hunk 2 above)
Comment (k-g-j) [reply ui/components/app/wallet-overview/wallet-overview.js:26]: Done!

File: ui/components/multichain/account-list-menu/account-list-menu.js [hunk 3]
```diff
@@ -64,6 +66,29 @@ const ACTION_MODES = {
   IMPORT: 'import',
 };
 
+///: BEGIN:ONLY_INCLUDE_IF(keyring-snaps)
+const mergeAccounts = (accounts, internalAccounts) => {
```
Comment (owencraston) [inline ui/components/multichain/account-list-menu/account-list-menu.js:70]: nitpick: The name `accounts` is not very descriptive and makes me wonder what this function is really doing? Is there a more descriptive name we can use for these params? I would also really love if you added a short jsdoc to this fucntion.

File: ui/helpers/utils/accounts.js [hunk 4]
```diff
@@ -60,3 +66,27 @@ export function getAvatarNetworkColor(name) {
       return undefined;
   }
 }
+
```
Comment (owencraston) [inline ui/helpers/utils/accounts.js:69]: how hard would it be to convert this file to typescript? If its not a major change can we do that here before this file grows too big?

File: ui/components/multichain/account-list-item/account-list-item.js [hunk 5]
```diff
@@ -57,27 +53,6 @@ import { useAccountTotalFiatBalance } from '../../../hooks/useAccountTotalFiatBa
 const MAXIMUM_CURRENCY_DECIMALS = 3;
 const MAXIMUM_CHARACTERS_WITHOUT_TOOLTIP = 17;
 
-function getLabel(t, { type }) {
```
Comment (owencraston) [inline ui/components/multichain/account-list-item/account-list-item.js:60]: This method has been replaced by `getAccountLabel`. I only see `getAccountLabel` gets called inside a code fence but the `getLabel` account was not fenced. Does this mean we are missing a use case for `getAccountLabel`?

File: ui/components/multichain/account-list-menu/account-list-menu.js (hunk 3 above)
Comment (k-g-j) [reply ui/components/multichain/account-list-menu/account-list-menu.js:70]: Done!

File: ui/helpers/utils/accounts.js (hunk 4 above)
Comment (k-g-j) [reply ui/helpers/utils/accounts.js:69]: Good idea but probably now feasible in this pr

File: ui/components/multichain/account-list-item/account-list-item.js (hunk 5 above)
Comment (k-g-j) [reply ui/components/multichain/account-list-item/account-list-item.js:60]: This is a good point and I believe we can remove the code fencing around the larger function

File: ui/helpers/utils/accounts.js [hunk 6]
```diff
@@ -60,3 +66,32 @@ export function getAvatarNetworkColor(name) {
...
+export function getAccountLabel(type, account) {
```
Comment (gantunesr) [inline ui/helpers/utils/accounts.js:70]: Lets add some unit tests here

File: ui/components/multichain/account-list-menu/account-list-menu.js [hunk 7]
```diff
@@ -64,6 +66,34 @@ const ACTION_MODES = {
...
+        label: getAccountLabel(
+          internalAccount.metadata.keyring.type,
+          internalAccount,
+        ),
+      };
+    }
+    return account;
+  });
+};
+
```
Comment (gantunesr) [inline ui/components/multichain/account-list-menu/account-list-menu.js:96]: This change seems unrelated to the PR, where is it coming from?

File: ui/components/app/wallet-overview/wallet-overview.js (hunk 2 above)
Comment (gantunesr) [inline ui/components/app/wallet-overview/wallet-overview.js:26]: This way is safer, in case we want to make a change we just need to change the internal logic of `getAccountLabel`

```suggestion
  const { keyring } = selectedAccount.metadata;
  const label = selectedAccount.metadata.snap?.name
  ? getAccountLabel(keyring.type, account)
  : null;
```

File: ui/components/app/wallet-overview/wallet-overview.js [hunk 8]
```diff
@@ -24,6 +33,16 @@ const WalletOverview = ({
...
         ) : null}
         {balance}
+        {label ? (
+          <Tag
+            label={label}
+            labelProps={{
+              variant: TextVariant.bodyXs,
+              color: Color.textAlternative,
+            }}
+            startIconName={keyring === KeyringType.snap ? IconName.Snaps : null}
```
Comment (gantunesr) [inline ui/components/app/wallet-overview/wallet-overview.js:43]: Based on the previous suggestion, this should be update like this
```suggestion
            startIconName={keyring.type === KeyringType.snap ? IconName.Snaps : null}
```

File: ui/helpers/utils/accounts.js (hunk 6 above)
Comment (k-g-j) [reply ui/helpers/utils/accounts.js:70]: Done :)

File: ui/components/multichain/account-list-menu/account-list-menu.js (hunk 7 above)
Comment (k-g-j) [reply ui/components/multichain/account-list-menu/account-list-menu.js:96]: It is needed because we have not replaced MetaMask Identities with internal accounts
so when the component calls `const accounts = useSelector(getMetaMaskAccountsOrdered);` to get a list of accounts those accounts do not have the data that internalAccounts do BUT they have balances which internal accounts do not have. So we have to merge the results of that selector with the data we also need from the internal accounts in order to have all the data WITH the balances

Comment (gantunesr) [reply ui/components/multichain/account-list-menu/account-list-menu.js:96]: Ah ok, that will be a little bit of technical debt, lets remember to resolve with the rest of the accounts-controller changes

**Review by owencraston — APPROVED** [review]
Nice work! 🥳 
Screenshot 2023-12-19 at 6 50 12 PM

**Review by gantunesr — APPROVED** [review]
LGTM. The feature works as expected. Ship it

https://github.com/MetaMask/metamask-extension/assets/17601467/d18dec30-0d23-47e4-8c71-ef14f9e1ffff
